crates/jit/src/link.rs

```diff
@@ -1,33 +1,48 @@
 //! Linking for JIT-compiled code.
 
-use crate::Compilation;
+use crate::CodeMemory;
 use cranelift_codegen::binemit::Reloc;
+use cranelift_codegen::ir::JumpTableOffsets;
 use std::ptr::{read_unaligned, write_unaligned};
+use wasmtime_environ::entity::PrimaryMap;
+use wasmtime_environ::wasm::DefinedFuncIndex;
 use wasmtime_environ::{Module, Relocation, RelocationTarget};
 use wasmtime_runtime::libcalls;
 use wasmtime_runtime::VMFunctionBody;
 
 /// Links a module that has been compiled with `compiled_module` in `wasmtime-environ`.
 ///
 /// Performs all required relocations inside the function code, provided the necessary metadata.
-pub fn link_module(module: &Module, compilation: &Compilation) {
-    for (fatptr, r) in compilation.code_memory.unpublished_relocations() {
+pub fn link_module<'a>(
```

Comment:
Is the goal here to largely get rid of the `Compilation` structure? I used that for a bit as a convenient "here's a package of a whole bunch of stuff" so when a field was added you didn't have to go add a bunch of parameters to a bunch of functions.

Also I think the `<'a>` here can be removed?

Reply:
yes, the caller "shreds" fields of the `Compilation` and changes their ownership -- only few are useful here (see `LinkContext` below). Passing entire `Compilation` into `link_module` blocks few compilation operations to be done sooner.